### DISCUSSION TITLE
Access Denied on Windows

### DISCUSSION THREAD

**Discussion by seikosantana** [scottoffen/grapevine#85] [Q&A] [answered]
Hi, I am trying to start a server on a specified port by clearing and adding prefix, but I got Access Denied when invoking server.Run(). the exception does not occur if i commented out the prefix clearing and adding prefix part. What is the cause of Access Denied? Did i miss anything?

Thanks in advace.

**Comment by scottoffen** [comment]
Can you post some code?

**Reply by seikosantana** [reply]
Sure.
```cs
using System;
using Grapevine;
using System.IO;
namespace HttpServer
{
    class Program
    {
        public static void Main(string[] args)
        {
            Run();
            Console.WriteLine("Built");
        }

        public readonly static string configPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "config.json");

        public static void Run()
        {
            var server = RestServerBuilder.UseDefaults().Build();
            server.Prefixes.Clear();
            server.Prefixes.Add("http://+:1234/");
            var materialsRoute = new Route(async (context) =>
            {
                await context.Response.SendResponseAsync(File.ReadAllText(configPath));
            }, "Get", "/materials");
            server.Router.Register(materialsRoute);
            server.Run();
        }
    }
}
```

The access denied exception is thrown on the `server.Run()`.

**Reply by scottoffen** [reply]
Do you have something else on your machine that is using port `1234`?

**Reply by seikosantana** [reply]
No, there isn't anything running on port 1234. I tried changing it to any port, be it 1234, 5555, 8888, 9999, i got access denied.

**Reply by seikosantana** [reply]
Current workaround is to run Visual Studio as administrator.

**Comment by scottoffen** [comment, marked as answer]
## Access Denied ##

When you try to run your server for the first time, you might see an Access Denied exception occur. While annoying, this is rather easy to overcome. If you want to use HTTP.SYS to listen on a URI, you either need administrative privileges or an administrator that reserves the URI for normal users.

### Run As Administrator ###

For development purposes, assuming you are an administrator on your Widows instance, you can simply run Visual Studio as administrator. Likewise, you can run the Windows command line as administrator if you want to test it outside of Visual Studio.

### Configure Namespace Reservations ###

For obvious reasons, counting on your application running as administrator in production is unwise. In these instances, you'll want to configure a namespace reservation.

This reservation can be made by using the [`netsh`](https://msdn.microsoft.com/en-us/library/windows/desktop/cc307236%28v=vs.85%29.aspx) utility.

```cmd
$ netsh http add urlacl url=http://+:1234/ user=DOMAIN\user
```

This can be done programatically, but the implementation is outside the scope of Grapevine. It's important that the value of the `url` argument is an exact match, or the exception will still be thrown.

```csharp
server.Prefixes.Add("http://+:1234/");
```

See the documentation for the [`add urlacl`](https://msdn.microsoft.com/en-us/library/windows/desktop/cc307223%28v=vs.85%29.aspx) for a description of the parameters.

**Comment by magnuso68** [comment]
Hi, first of all thanks for a wonderful piece of software.
Worked as a dream to set up some routes for my Winform app.

But the advice in this thread is not working for me. I have carefully added with `netsh http add urlacl url=http://+:1234/ user=DOMAIN\user` checked with `netsh http add urlacl show` that it has been registred but still I got an error when I start the server after using `server.Prefixes.Add("http://+:1234/");`. 

The only solution to be able to run my application is to start it as administrator. Kind of annoying.
Thanks
